src/reader.rs

```diff
@@ -4524,95 +4536,72 @@
 }
 
 impl<'a> Reader<'a> {
-    fn exec_mode_prompt(&mut self) {
-        self.mode_prompt_buff.clear();
-        if function::exists(MODE_PROMPT_FUNCTION_NAME, self.parser) {
-            let mut mode_indicator_list = vec![];
-            exec_subshell(
-                MODE_PROMPT_FUNCTION_NAME,
-                self.parser,
-                Some(&mut mode_indicator_list),
-                false,
-            );
-            // We do not support multiple lines in the mode indicator, so just concatenate all of
-            // them.
-            for i in mode_indicator_list {
-                self.mode_prompt_buff.push_utfstr(&i);
-            }
-        }
-    }
-
-    /// Reexecute the prompt command. The output is inserted into prompt_buff.
-    fn exec_prompt(&mut self) {
-        // Clear existing prompts.
-        self.left_prompt_buff.clear();
-        self.right_prompt_buff.clear();
-
+    /// Execute prompt commands based on the provided arguments. The output is inserted into prompt_buff.
+    fn exec_prompt(&mut self, exec_left: bool, exec_right: bool) {
         // Suppress fish_trace while in the prompt.
+        // Prompts must be run non-interactively.
         let mut zelf = scoped_push_replacer_ctx(
             self,
             |zelf, new_value| {
-                std::mem::replace(
-                    &mut zelf.parser.libdata_mut().suppress_fish_trace,
-                    new_value,
+                let mut libdata = zelf.parser.libdata_mut();
+                (
+                    std::mem::replace(&mut libdata.suppress_fish_trace, new_value.0),
+                    std::mem::replace(&mut libdata.is_interactive, new_value.1),
                 )
             },
-            true,
+            (true, false),
         );
 
         // Update the termsize now.
         // This allows prompts to react to $COLUMNS.
         zelf.update_termsize();
 
-        // If we have any prompts, they must be run non-interactively.
-        if !zelf.conf.left_prompt_cmd.is_empty() || !zelf.conf.right_prompt_cmd.is_empty() {
-            let mut zelf = scoped_push_replacer_ctx(
-                &mut zelf,
-                |zelf, new_value| {
-                    std::mem::replace(&mut zelf.parser.libdata_mut().is_interactive, new_value)
+        zelf.mode_prompt_buff.clear();
+        if function::exists(MODE_PROMPT_FUNCTION_NAME, zelf.parser) {
+            let mut mode_indicator_list = vec![];
+            exec_subshell(
+                MODE_PROMPT_FUNCTION_NAME,
+                zelf.parser,
+                Some(&mut mode_indicator_list),
+                false,
+            );
+            // We do not support multiline mode indicators, so just concatenate all of them.
+            zelf.mode_prompt_buff = WString::from_iter(mode_indicator_list);
```

Comment:
Right, this is how things used to work before 8ff866b26b (Add repaint-mode bind function, 2019-04-01),
and the optimization added by that commit will still work.

Actually before that commit we didn't execute the mode prompt if both left and right prompts were empty.
But that seems like an accident from 767742c7e7 (Rework how the mode is reported in fish_vi_mode, 2015-06-04).